### PR TITLE
[Homo Optimization] Optimization on GSF

### PR DESCRIPTION
**Pull request by jalencato**
*Issue #, if available:*

*Description of changes:*

Allow users to start training/inferring job without specifying target_ntype/target_etype on homogeneous graph.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/configuration/configuration-run.rst [hunk 1]
```diff
@@ -381,20 +381,20 @@ Classification and Regression Task
 
 Node Classification/Regression Specific
 .........................................
-- **target_ntype**: (**Required**) The node type for prediction.
+- **target_ntype**: The node type for prediction.
 
     - Yaml: ``target_ntype: movie``
     - Argument: ``--target-ntype movie``
-    - Default value: This parameter must be provided by user.
+    - Default value: This parameter must be provided by user. If not provided, will be default to "_N", and require homogeneous graph input.
```
Comment (classicsong) [inline docs/source/configuration/configuration-run.rst:388]:
```suggestion
    - Default value: For heterogeneous input graph, this parameter must be provided by the user. If not provided, GraphStorm will assume the input graph is a homogeneous graph and set ``target_ntype`` to "_N".
```

File: docs/source/configuration/configuration-run.rst [hunk 2]
```diff
@@ -381,20 +381,20 @@ Classification and Regression Task
...
+- **target_etype**: The list of canonical edge types that will be added as a training target in edge classification/regression tasks, for example ``--train-etype query,clicks,asin`` or ``--train-etype query,clicks,asin query,search,asin``. A canonical edge type should be formatted as `src_node_type,relation_type,dst_node_type`. Currently, GraphStorm only supports single task edge classification/regression, i.e., it only accepts one canonical edge type.
 
     - Yaml: ``target_etype:``
            | ``- query,clicks,asin``
     - Argument: ``--target-etype query,clicks,asin``
-    - Default value: This parameter must be provided by user.
+    - Default value: This parameter must be provided by user. If not provided, will be default to ("_N", "_E", "_N"), and require homogeneous graph input.
```
Comment (classicsong) [inline docs/source/configuration/configuration-run.rst:397]: Ditto

File: python/graphstorm/config/argument.py [hunk 3]
```diff
@@ -1573,9 +1574,12 @@ def target_ntype(self):
...
         """
         # pylint: disable=no-member
-        assert hasattr(self, "_target_ntype"), \
-            "Must provide the target ntype through target_ntype"
-        return self._target_ntype
+        if hasattr(self, "_target_ntype"):
+            return self._target_ntype
+        else:
+            logging.warning("There is not target ntype provided, "
+                            "will treat as homogeneous graph")
```
Comment (classicsong) [inline python/graphstorm/config/argument.py:1581]:
```suggestion
                            "will treat the input graph as a homogeneous graph")
```

File: python/graphstorm/config/argument.py [hunk 4]
```diff
@@ -1648,8 +1652,10 @@ def target_etype(self):
             classification/regression. Support multiple tasks when needed.
         """
         # pylint: disable=no-member
-        assert hasattr(self, "_target_etype"), \
-            "Edge classification task needs a target etype"
+        if not hasattr(self, "_target_etype"):
+            logging.warning("There is not target etype provided, "
+                            "will treat as homogeneous graph")
```
Comment (classicsong) [inline python/graphstorm/config/argument.py:1657]: Ditto

File: python/graphstorm/dataloading/dataset.py [hunk 5]
```diff
@@ -553,6 +554,13 @@ def __init__(self, graph_name, part_config, train_etypes, eval_etypes=None,
                                                  decoder_edge_feat,
                                                  lm_feat_ntypes=lm_feat_ntypes,
                                                  lm_feat_etypes=lm_feat_etypes)
+        if self._train_etypes == [DEFAULT_ETYPE]:
+            assert self._g.ntypes == [DEFAULT_NTYPE] and \
+                   self._g.etypes == [DEFAULT_ETYPE[1]], \
+                f"It is required to be a homogeneous graph when not providing " \
+                f"target_etype on node task, expect node type {[DEFAULT_NTYPE]} and " \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:561]: Why node task needs target_etype?

Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:561]:
```suggestion
                f"target_etype on node task, expect node type to be {[DEFAULT_NTYPE]} and " \
```

File: python/graphstorm/dataloading/dataset.py [hunk 6]
```diff
@@ -553,6 +554,13 @@ def __init__(self, graph_name, part_config, train_etypes, eval_etypes=None,
...
+                f"edge type {[DEFAULT_ETYPE[1]]}, but get {self._g.ntypes} " \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:562]:
```suggestion
                f"edge type to be {[DEFAULT_ETYPE[1]]}, but get {self._g.ntypes} " \
```

File: python/graphstorm/dataloading/dataset.py [hunk 7]
```diff
@@ -553,6 +554,13 @@ def __init__(self, graph_name, part_config, train_etypes, eval_etypes=None,
                                                  decoder_edge_feat,
                                                  lm_feat_ntypes=lm_feat_ntypes,
                                                  lm_feat_etypes=lm_feat_etypes)
+        if self._train_etypes == [DEFAULT_ETYPE]:
+            assert self._g.ntypes == [DEFAULT_NTYPE] and \
+                   self._g.etypes == [DEFAULT_ETYPE[1]], \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:559]: Should not DGL etypes be canonical etypes?

File: python/graphstorm/dataloading/dataset.py [hunk 8]
```diff
@@ -731,6 +739,13 @@ def __init__(self, graph_name, part_config, eval_etypes,
                                                  decoder_edge_feat,
                                                  lm_feat_ntypes=lm_feat_ntypes,
                                                  lm_feat_etypes=lm_feat_etypes)
+        if self._eval_etypes == [DEFAULT_ETYPE]:
+            assert self._g.ntypes == [DEFAULT_NTYPE] and \
+                   self._g.etypes == [DEFAULT_ETYPE[1]], \
+                f"It is required to be a homogeneous graph when not providing " \
+                f"target_etype on node task, expect node type {[DEFAULT_NTYPE]} and " \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:746]: ditto

File: tests/unit-tests/test_dataloading.py [hunk 9]
```diff
@@ -1354,6 +1355,92 @@ def test_inbatch_joint_neg_sampler(num_pos, num_neg):
...
+    th.distributed.init_process_group(backend='gloo',
+                                      init_method='tcp://127.0.0.1:23456',
+                                      rank=0,
+                                      world_size=1)
+    tr_ntypes = ["_N"]
+    va_ntypes = ["_N"]
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # get the test dummy distributed graph
+        dist_graph, part_config = generate_dummy_dist_graph(graph_name='dummy',
```
Comment (classicsong) [inline tests/unit-tests/test_dataloading.py:1369]: Is the dummpy dist graph a heterogeneous graph?
Can you add some comments here how the exceptions are triggerred?

File: python/graphstorm/dataloading/dataset.py (hunk 7 above)
Comment (jalencato) [reply python/graphstorm/dataloading/dataset.py:559]: Nope. For DGL graph, the input is just [DEFAULT_ETYPE[1]]

File: python/graphstorm/dataloading/dataset.py (hunk 5 above)
Comment (jalencato) [reply python/graphstorm/dataloading/dataset.py:561]: For homogeneous graph, it is strictly required for both of the node type and the edge type, so it is required to check both node type and edge type meet the requirement.

Comment (jalencato) [reply python/graphstorm/dataloading/dataset.py:561]: Meanwhile, code here is for GSgnnEdgeTrainData.

File: tests/unit-tests/test_dataloading.py (hunk 9 above)
Comment (jalencato) [reply tests/unit-tests/test_dataloading.py:1370]: Refer to the comments here about what each test case is and why it should trigger an error.

File: python/graphstorm/dataloading/dataset.py (hunk 7 above)
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:559]: Can you add a comment here?

File: python/graphstorm/dataloading/dataset.py [hunk 10]
```diff
@@ -553,6 +554,13 @@ def __init__(self, graph_name, part_config, train_etypes, eval_etypes=None,
                                                  decoder_edge_feat,
                                                  lm_feat_ntypes=lm_feat_ntypes,
                                                  lm_feat_etypes=lm_feat_etypes)
+        if self._train_etypes == [DEFAULT_ETYPE]:
+            assert self._g.ntypes == [DEFAULT_NTYPE] and \
+                   self._g.etypes == [DEFAULT_ETYPE[1]], \
+                f"It is required to be a homogeneous graph when not providing " \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:560]:
```suggestion
                f"It is required to be a homogeneous graph when target_etype is not provided or is set to {DEFAULT_ETYPE} on edge tasks" \
```

File: python/graphstorm/dataloading/dataset.py [hunk 11]
```diff
@@ -731,6 +739,13 @@ def __init__(self, graph_name, part_config, eval_etypes,
                                                  decoder_edge_feat,
                                                  lm_feat_ntypes=lm_feat_ntypes,
                                                  lm_feat_etypes=lm_feat_etypes)
+        if self._eval_etypes == [DEFAULT_ETYPE]:
+            assert self._g.ntypes == [DEFAULT_NTYPE] and \
+                   self._g.etypes == [DEFAULT_ETYPE[1]], \
+                f"It is required to be a homogeneous graph when not providing " \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:745]:
```suggestion
                f"It is required to be a homogeneous graph when target_etype is not provided or is set to {DEFAULT_ETYPE} on edge tasks " \
```

File: python/graphstorm/dataloading/dataset.py [hunk 12]
```diff
@@ -932,6 +946,13 @@ def __init__(self, graph_name, part_config, train_ntypes, eval_ntypes=None,
                                                  edge_feat_field=edge_feat_field,
                                                  lm_feat_ntypes=lm_feat_ntypes,
                                                  lm_feat_etypes=lm_feat_etypes)
+        if self._train_ntypes == [DEFAULT_NTYPE]:
+            assert self._g.ntypes == [DEFAULT_NTYPE] and \
+                   self._g.etypes == [DEFAULT_ETYPE[1]], \
+                f"It is required to be a homogeneous graph when not providing " \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:952]:
```suggestion
                f"It is required to be a homogeneous graph when target_ntype is not provided or is set to {DEFAULT_NTYPE} on node tasks" \
```

File: python/graphstorm/dataloading/dataset.py [hunk 13]
```diff
@@ -1072,6 +1093,13 @@ def __init__(self, graph_name, part_config, eval_ntypes,
                                                  lm_feat_ntypes=lm_feat_ntypes,
                                                  lm_feat_etypes=lm_feat_etypes)
 
+        if self._eval_ntypes == [DEFAULT_NTYPE]:
+            assert self._g.ntypes == [DEFAULT_NTYPE] and \
+                   self._g.etypes == [DEFAULT_ETYPE[1]], \
+                f"It is required to be a homogeneous graph when not providing " \
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:1099]:
```suggestion
                f"It is required to be a homogeneous graph when target_ntype is not provided or is set to {DEFAULT_NTYPE} on node tasks" \
```

File: tests/unit-tests/data_utils.py [hunk 14]
```diff
@@ -343,6 +343,83 @@ def generate_dummy_homo_graph(size='tiny', gen_mask=True):
 
     return hetero_graph
 
+def generate_dummy_homo_failure_graph(size='tiny', gen_mask=True, type='node'):
+    """
+    generate a dummy homogeneous graph.
```
Comment (classicsong) [inline tests/unit-tests/data_utils.py:348]: Can you add more comments about what is the purpose of this dummy homo failure graph?

File: tests/unit-tests/data_utils.py [hunk 15]
```diff
@@ -343,6 +343,83 @@ def generate_dummy_homo_graph(size='tiny', gen_mask=True):
 
     return hetero_graph
 
+def generate_dummy_homo_failure_graph(size='tiny', gen_mask=True, type='node'):
```
Comment (classicsong) [inline tests/unit-tests/data_utils.py:346]:
```suggestion
def generate_dummy_homogeneous_failure_graph(size='tiny', gen_mask=True, type='node'):
```

File: tests/unit-tests/data_utils.py [hunk 16]
```diff
@@ -431,6 +508,28 @@ def generate_dummy_dist_graph_multi_target_ntypes(dirname, size='tiny', graph_na
     return partion_and_load_distributed_graph(hetero_graph=hetero_graph, dirname=dirname,
                                               graph_name=graph_name)
 
+
+def generate_dummy_dist_graph_homo_failure_graph(dirname, size='tiny', graph_name='dummy',
```
Comment (classicsong) [inline tests/unit-tests/data_utils.py:512]:
```suggestion
def generate_dummy_dist_graph_homogeneous_failure_graph(dirname, size='tiny', graph_name='dummy',
```

**Review by classicsong — APPROVED** [review]

**Pull request merged**